src/uri_ext.rs

```diff
@@ -0,0 +1,98 @@
+use lsp_types::Uri;
+use std::borrow::Cow;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+#[cfg(not(windows))]
+pub use std::fs::canonicalize as strict_canonicalize;
+
+/// On Windows, rewrites the wide path prefix `\\?\C:` to `C:`  
+/// Source: https://stackoverflow.com/a/70970317
+#[inline]
+#[cfg(windows)]
+fn strict_canonicalize<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+    use std::io;
+
+    fn impl_(path: PathBuf) -> std::io::Result<PathBuf> {
+        let head = path
+            .components()
+            .next()
+            .ok_or(io::Error::new(io::ErrorKind::Other, "empty path"))?;
+        let disk_;
+        let head = if let std::path::Component::Prefix(prefix) = head {
+            if let std::path::Prefix::VerbatimDisk(disk) = prefix.kind() {
+                disk_ = format!("{}:", disk as char);
+                Path::new(&disk_).components().next().ok_or(io::Error::new(
+                    io::ErrorKind::Other,
+                    "failed to parse disk component",
+                ))?
+            } else {
+                head
+            }
+        } else {
+            head
+        };
+        Ok(std::iter::once(head)
+            .chain(path.components().skip(1))
+            .collect())
+    }
+
+    let canon = std::fs::canonicalize(path)?;
+    impl_(canon)
+}
+
+/// Provide more methods to [`fluent_uri::Uri`]. Especially to convert to and
+/// from file paths.
+pub trait UriExt {
+    /// Convert the path component of a [`fluent_uri::Uri`] to a file path.
+    fn to_file_path(&self) -> Option<Cow<Path>>;
+}
+
+impl UriExt for lsp_types::Uri {
+    fn to_file_path(&self) -> Option<Cow<Path>> {
```

Comment:
IIRC the `url` crate used to return a Result here (please double-check), so might be better to change to that as well

Reply:
It does return a `Result<Path, ()>`. I think that staying with `Option` is ok here.